adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/MqttContext.java

```diff
@@ -258,6 +258,18 @@ public ResourceIdentifier topic() {
         return topic;
     }
 
+    /**
+     * Sets the topic of this message.
+     * <p>
+     * This overwrites the topic determined from the message.
+     *
+     * @param topic The topic.
+     * @throws NullPointerException if topic is {@code null}.
+     */
+    public void setTopic(final ResourceIdentifier topic) {
```

Comment:
AFAIC we should not allow changing the topic once the context has been created from a message (and the topic it has been published to). I am concerned that if we do open up this door, we will end up with handling special cases all over the place. Currently, our concept of encoding all relevant information in the topic is sound and tested. I wouldn't want to give it up that easily.
IMHO we will need to take another look at the mapping mechanism itself and maybe fix it at a conceptual level to make it compatible with the current MQTT message handling. WDYT @calohmn @kaniyan ?

Reply:
I can imagine to add something like an *applyMappedTargetAddress(ResourceIdentifier)* method to *MqttContext* which would then update the context's state based on the given target address. This way we would at least define a well defined semantic context for updating the device ID within the MqttContext as opposed to simply letting everybody alter the original MQTT message's topic without really knowing what this entails.

Reply:
@BobClaerhout thoughts?

Reply:
I'm not too familiar with the context but I can imagine you would want to keep this as closed as possible. I was already wondering whether  adding this method would be allowed.  
Nevertheless I think updating the concept in order to have the mappedDeviceId to determine whether it's a fatal error, will be quite disruptive. Applying the mappedTargetAddress is indeed an option by means of another method  name instead of updating the topic directly. Another option might be to add a separate topic/targetAddress called `mappedTargetAddress` which doesn't really change the actual topic but is kept separate. An additional method can be something like `getMappedTargetAddressOrDefault` which returns the mapped target address and if it doesn't exists falls back to the default topic. Only question is when to use which. Just thinking out loud.

Reply:
I have taken a more thorough look at how we are currently using MqttContext and it seems that encapsulation is currently not being imposed to the degree that I would personally like to see. I suppose that we will need to review the MqttContext class' usage in its entirety in order to come to a better solution. However, we would really like to do the 1.11.0 release during the next few days and so IMHO it would be ok to add an *applyMappedTargetAddress(ResourceIdentifier)* method for the time being and then review the whole thing as part of 1.12.0. Would that work for you @BobClaerhout? Note that this will probably mean changing some code around the mapping in 1.12.0.
@calohmn @kaniyan would that be ok for you too?

Reply:
I was a bit surprised about this feature of overriding the device id, I didn't know about it.
Looking at the implementation, and also at this PR, I don't see any practical issues.

But I agree that updating the topic in the `MqttContext` doesn't look that good.
This would also mean that `ctx.topic` is then possibly fundamentally different from `ctx.message().topicName()`.
(Both are already potentially different since the latter might contain a propertyBag at the end while the former doesn't, but at least the path is always the same.)

@BobClaerhout I also think that adding a `MqttContext#mappedTargetAddress` field with corresponding setter would be a good alternative to updating the `MqttContext` topic.
The `MqttContext#tenant()` and `MqttContext#deviceId()` methods need to be updated then to derive their value from the `mappedTargetAddress` first (if set).
A `getMappedTargetAddressOrDefault()` method would then not be really needed. I've checked the invocations of `MqttContext#topic()`, it's only done before the message mapping - with one exception: the `ErrorSubscription#getErrorPublishTopic()` method. This method should be changed to get `contextTopicDeviceId` from `MqttContext#deviceId()`.

Reply:
@sophokles73 Hadn't seen your comment yet when writing my comment above.

So, I would be fine with adding a `setMappedTargetAddress(ResourceIdentifier)` or `applyMappedTargetAddress(ResourceIdentifier)` method in `MqttContext` now for 1.11.0 and then review/improve the whole mapping integration for 1.12.0.

Reply:
I'm fine with the other method name. Updating the mapping code in 1.12.0 is also fine for me. I already updated it in the PR. Thanks for the flexibility, greatly appreciated that we can add a quick fix in 1.11.0 already.

Reply:
> A getMappedTargetAddressOrDefault() method would then not be really needed. I've checked the invocations of MqttContext#topic(), it's only done before the message mapping - with one exception: the ErrorSubscription#getErrorPublishTopic() method. This method should be changed to get contextTopicDeviceId from MqttContext#deviceId().

Or, maybe even better, the *ErrorSubscription#getErrorPublishTopic()* method could be moved into the MqttContext class altogether. This would also increase encapsulation ...

Reply:
I think one main aspect to the Hono 1.12.0 improvements would be to make it clear in the code that the mapping mechanism can only override the device id, not the tenant id.
As of now, this is only implicit in the HTTP mapping implementation:
https://github.com/eclipse/hono/blob/cedb10966db42f8473d1cb56d3daee88f3eede6d/adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java#L268-L269